### PR TITLE
Support identifiers with `.` in them

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1432

 # Rationale for this change
Field names containing period such as `f.c1` cannot be named in SQL query

# What changes are included in this PR?
1. Directly construct `Expr::Column` references rather than trying to parse a string using `col()` when converting SQL

# Are there any user-facing changes?
columns with periods can now be used

### CODE REVIEW

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -1323,9 +1325,14 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
                     Ok(Expr::ScalarVariable(var_names))
                 } else {
-                    // create a column expression based on raw user input, this column will be
-                    // normalized with qualifer later by the SQL planner.
-                    Ok(col(&id.value))
+                    // Don't use `col()` here because it will try to
+                    // interpret names with '.' as if they were
+                    // compound indenfiers, but this is not a compound
+                    // identifier. (e.g. it is "foo.bar" not foo.bar)
+                    Ok(Expr::Column(Column {
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:1332]: This is the core change -- `col()` attempts to parse a string as a potentially compound identifier. For a `SQLExpr::Identifier` this should not be done and the value should be used as an unqualified relation

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -1062,8 +1062,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 }
 
                 let field = schema.field(field_index - 1);
-                let col_ident = SQLExpr::Identifier(Ident::new(field.qualified_name()));
-                self.sql_expr_to_logical_expr(&col_ident, schema)?
+                Expr::Column(field.qualified_column())
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:1065]: This change avoids trying to reparse the string as a column name, and instead just creates the `Column` reference directly

**Comment by alamb** [comment]
FYI @liukun4515

**Review by houqp — APPROVED** [review]
Simple and elegant fix, thanks @alamb !

**Review by Jimexist — APPROVED** [review]

File: datafusion/tests/sql.rs [hunk 3]
```diff
@@ -5426,6 +5426,67 @@ async fn qualified_table_references() -> Result<()> {
...
+        error.to_string(),
+        "No field named 'f1.c1'. Valid fields are 'test.f.c1', 'test.test.c2'"
+    );
+
+    // however, enclosing it in double quotes is ok
+    let sql = format!(r#"SELECT "f.c1" from test"#);
+    let actual = execute_to_batches(&mut ctx, &sql).await;
+    let expected = vec![
+        "+--------+",
+        "| f.c1   |",
```
Comment (Jimexist) [inline datafusion/tests/sql.rs:5461]: i wonder if it is okay to create column `....`

**Review by liukun4515 — APPROVED** [review]
LGTM

File: datafusion/tests/sql.rs [hunk 4]
```diff
@@ -5426,6 +5426,67 @@ async fn qualified_table_references() -> Result<()> {
...
+        ("f.c1", Arc::new(c1) as ArrayRef),
+        //  evil -- use the same name as the table
+        ("test.c2", Arc::new(c2) as ArrayRef),
+    ])?;
+
+    let table = MemTable::try_new(batch.schema(), vec![vec![batch]])?;
+    ctx.register_table("test", Arc::new(table))?;
+
+    // referring to the unquoted column is an error
+    let sql = format!(r#"SELECT f1.c1 from test"#);
```
Comment (xudong963) [inline datafusion/tests/sql.rs:5449]: Clippy: useless_format

File: datafusion/src/sql/planner.rs (hunk 2 above)
Comment (xudong963) [reply datafusion/src/sql/planner.rs:1065]: 👍, FYI @hntd187

**Review by xudong963 — APPROVED** [review]
LGTM, thanks @alamb

File: datafusion/src/sql/planner.rs (hunk 2 above)
Comment (hntd187) [reply datafusion/src/sql/planner.rs:1065]: This is much better than what I had :-)

Comment (alamb) [reply datafusion/src/sql/planner.rs:1065]: We are all learning together!

File: datafusion/tests/sql.rs (hunk 4 above)
Comment (alamb) [reply datafusion/tests/sql.rs:5449]: will fix

File: datafusion/tests/sql.rs (hunk 3 above)
Comment (alamb) [reply datafusion/tests/sql.rs:5464]: I think it is "ok" from the point of view of DataFusion should run the query. I don't think anyone should actually do it 😆 

Added a test for this case in 3e4418c67

**Pull request merged**